### PR TITLE
xdg-surface: for each surface

### PR DESCRIPTION
**Pull request by acrisci**
Implement `wlr_xdg_surface_v6_for_each_surface`.

### CODE REVIEW

File: src/types/shell/xdg_shell_v6.rs [hunk 1]
```diff
@@ -181,6 +183,21 @@ impl XdgV6ShellSurface {
...
+    pub fn for_each_surface(&self, iterator: &FnMut(SurfaceHandle, i32, i32)) {
+        unsafe {
+            extern fn c_iterator(wlr_surface: *mut wlr_surface, sx: i32, sy: i32, data: *mut c_void) {
+                unsafe {
+                    let iterator = &mut *(data as *mut Box<&Fn(SurfaceHandle, i32, i32)>);
+                    let surface = SurfaceHandle::from_ptr(wlr_surface);
+                    iterator(surface, sx, sy);
+                }
+            }
+            let mut iterator = Box::new(iterator);
```
Comment (Timidger) [inline src/types/shell/xdg_shell_v6.rs]: Does this need to be boxed? Can you not just pass the `&FnMut` as a function pointer to the function?

**Comment by acrisci** [comment]
https://stackoverflow.com/questions/24191249/working-with-c-void-in-an-ffi

File: src/types/shell/xdg_shell_v6.rs [hunk 2]
```diff
@@ -181,6 +183,20 @@ impl XdgV6ShellSurface {
         }
     }
 
+    pub fn for_each_surface(&self, mut iterator: &FnMut(SurfaceHandle, i32, i32)) {
+        unsafe {
+            extern fn c_iterator(wlr_surface: *mut wlr_surface, sx: i32, sy: i32, data: *mut c_void) {
+                unsafe {
+                    let iterator = &mut *(data as *mut &mut FnMut(SurfaceHandle, i32, i32));
```
Comment (Timidger) [inline src/types/shell/xdg_shell_v6.rs]: You are casting an `*mut & FnMut(SurfaceHandle, i32, i32)` into an *mut &mut FnMut(SurfaceHandle, i32, i32). You should make it take an `&mut FnMut`, or cast this line into an `&FnMut(...)`.

File: src/types/shell/xdg_shell_v6.rs [hunk 3]
```diff
@@ -181,6 +183,20 @@ impl XdgV6ShellSurface {
         }
     }
 
+    pub fn for_each_surface(&self, mut iterator: &FnMut(SurfaceHandle, i32, i32)) {
+        unsafe {
+            extern fn c_iterator(wlr_surface: *mut wlr_surface, sx: i32, sy: i32, data: *mut c_void) {
```
Comment (Timidger) [inline src/types/shell/xdg_shell_v6.rs]: Just do `unsafe extern "C" fn` here instead of having an unsafe block in there.

**Comment by acrisci** [comment]
I wish I could get rid of all those muts, because we don't really mutate the iterator, but when I can't call the function because it says I can't borrow it as mutable.

**Comment by Timidger** [comment]
Yeah unfortunately that is just a consequence of using the `FnMut` trait. In this case that function could do some mutation of the arguments, so all those `mut`s are necessary.

**Comment by Timidger** [comment]
Thanks!

**Pull request merged**